tool/lib/webqa-canonical-generator.js

```diff
@@ -97,9 +105,23 @@ class CanonicalGenerator {
                 if (!arg.metadata.canonical)
                     continue;
 
-                // copy base canonical if npp canonical is missing
-                if (arg.metadata.canonical.base && !arg.metadata.canonical.npp)
-                    arg.metadata.canonical.npp = [...arg.metadata.canonical.base];
+                // copy base canonical if property canonical is missing
+                if (arg.metadata.canonical.base && !arg.metadata.canonical.property)
+                    arg.metadata.canonical.property = [...arg.metadata.canonical.base];
+
+                // if property is missing, try to use entity type info
+                if (!('property' in arg.metadata.canonical)) {
+                    let typestr = entityTypeToString(query.getArgType(arg.name));
+                    // only apply this if the type is unique
+                    if (typestr && typeCounts[typestr] === 1) {
+                        let base = typestr.substring(typestr.indexOf(':') + 1)
+                            .replace(/_/, ' ')
```

Comment:
Should probably use "clean()" from Utils instead of this replace, as that handles different casing conventions and does a few more fixes.